Use ThreadSafeWeakHashSet for Thread

geoffreygaren opened the pull request:
#### 0f137772c935fe6f8d188ab40bceefe731c719e2
<pre>
Use ThreadSafeWeakHashSet for Thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=302743">https://bugs.webkit.org/show_bug.cgi?id=302743</a>
<a href="https://rdar.apple.com/165000250">rdar://165000250</a>

Reviewed by Marcus Plutowski.

Because SaferCPP objects to HashMap&lt;Thread*, ...&gt;.

Re-landed 258267@main now that 287313@main has fixed the bugs that originally
blocked the conversion to ThreadSafeWeakHashSet.

Updated clients to use Ref/RefPtr instead of std::shared_ptr.

Also changed ThreadSafeWeakHashSet and ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr
to use WordLock instead of Lock. Thread and ThreadGroup require a lock that can
be used without access to TLS.

Also adopted WTF_CAPABILITY_LOCK annotations for WordLock, so that we don&apos;t
regress thread-safety analysis coverage by switching to WordLock. This required
suppressing some existing inconsistencies (which are fine, just not verifiable).

Canonical link: <a href="https://commits.webkit.org/303280@main">https://commits.webkit.org/303280@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a36370c3e40e8d3d4efa95d01723728bb97d8696

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131863 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4355 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42905 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139375 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83749 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/08b970c8-6f36-426b-8ad1-c383bdd7c577) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133733 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4292 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4117 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100793 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83749 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e55312c0-9453-46fc-bead-6c019bc53526) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134809 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118132 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81583 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/402eefa9-b064-4e16-831e-d9a4105726bb) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2960 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82595 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/123926 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/111704 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/36256 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142018 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130370 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4024 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36835 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109168 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4105 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3523 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109334 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3063 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114412 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57241 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20507 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4078 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32812 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163336 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3910 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67525 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42466 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4170 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4038 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -178,7 +178,7 @@ class ThreadSafeWeakPtrControlBlock {
     void setStrongReferenceCountDuringInitialization(size_t count) WTF_IGNORES_THREAD_SAFETY_ANALYSIS { m_strongReferenceCount = count; }
 
-    mutable Lock m_lock;
+    mutable WordLock m_lock;
```
It might just be easier to use a single 64-bit atomic of the strong/weak reference count. AFAIK, ARMv7 supports 64-bit atomics lock free so it be possible on all ports.

geoffreygaren replied:
Yeah, I think something like that is probably the best long term direction. But I'd prefer not to do it in this patch.

kmiller68 replied:
Yeah, fair enough, I don't know how hard it would be to write so I thought I'd mention it.

Achierius reviewed (approved):
r=me

geoffreygaren committed:
Use ThreadSafeWeakHashSet for Thread
https://bugs.webkit.org/show_bug.cgi?id=302743
rdar://165000250

Reviewed by Marcus Plutowski.

Because SaferCPP objects to HashMap<Thread*, ...>.

Re-landed 258267@main now that 287313@main has fixed the bugs that originally
blocked the conversion to ThreadSafeWeakHashSet.

Updated clients to use Ref/RefPtr instead of std::shared_ptr.

Also changed ThreadSafeWeakHashSet and ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr
to use WordLock instead of Lock. Thread and ThreadGroup require a lock that can
be used without access to TLS.

Also adopted WTF_CAPABILITY_LOCK annotations for WordLock, so that we don't
regress thread-safety analysis coverage by switching to WordLock. This required
suppressing some existing inconsistencies (which are fine, just not verifiable).

Canonical link: https://commits.webkit.org/303280@main
